### PR TITLE
Signing bootstrap workflow

### PR DESCRIPTION
**Pull request by aweiteka**
Provide mechanism for automated sigstore discovery.

- [x] implement sigstore config (~~pending #594~~) 
- [x] implement drop in trust policy (~~pending #604~~)
- [ ] ~~add integration tests~~

Example output:
```
$ sudo atomic pull --storage docker docker.io/aweiteka/hello-world
Found registry sigstore metadata image docker.io/aweiteka/sigstore
ID: aweiteka@gmail.com
Fingerprint: B3B04F8CF186436EF8F1CDAD7C6ACC9EE3A31016
Public key download URL: https://pgp.mit.edu/pks/lookup?op=get&search=0xFD5EB4DB480717ED
Do you want to add trust policy for this registry? (y/N) y
TODO: Writing trust config for docker.io/aweiteka to /etc/containers/registries.d
Public key aweiteka@gmail.com already installed at /etc/pki/containers/aweiteka@gmail.com
TODO: Adding trust policy: docker.io/aweiteka /etc/pki/containers/aweiteka@gmail.com sigstore.example.com:8443
There was a problem configuring the trust policy
```

### CODE REVIEW

File: Atomic/pull.py [hunk 1]
```diff
@@ -29,5 +31,75 @@ def pull_image(self):
...
+        Prompt user for trust on first use workflow
+        :param labels: dict of metadata labels defining sigstore trust
+        :return: True if user accepts
+        """
+        write_out("ID: " + labels['pubkey-id'])
+        write_out("Fingerprint: " + labels['pubkey-fingerprint'])
+        write_out("Public key download URL: %s" % labels['pubkey-url'])
+        if self.args.trust:
+            return True
+        trust = raw_input("Do you want to add trust policy for this registry? (y/N) ")
```
Comment (aweiteka) [inline Atomic/pull.py]: @runcom `raw_input()` is failing python 3. I wasn't sure about our compatibility goals. Should I use `input()`?

File: Atomic/pull.py [hunk 2]
```diff
@@ -27,6 +29,9 @@ def pull_docker_image(self):
         tag = tag if tag != "" else "latest"
         fq_name = skopeo_inspect("docker://{}".format(self.args.image))['Name']
         image = "docker-daemon:{}:{}".format(fq_name, tag)
+        if get_atomic_config_item(['discover_sigstores'], get_atomic_config()):
+            if not self.discover_sigstore():
+                write_out("There was a problem configuring the trust policy")
```
Comment (baude) [inline Atomic/pull.py]: We usually raise something here, ValueError?

File: Atomic/pull.py [hunk 3]
```diff
@@ -41,3 +46,71 @@ def pull_image(self):
...
+    def _get_sigstore_image_metadata(self, registry, repo):
+        """
+        Get sigstore metadata image
+        :param registry: registry string
+        :param repo: repo string
+        :return True on success
+        """
+        _img = get_atomic_config_item(['sigstore_metadata_image'], get_atomic_config())
+        sigstoreimage = '/'.join([registry, repo, _img])
+        data = skopeo_inspect("docker://" + sigstoreimage, args=None, fail_silent=True)
```
Comment (baude) [inline Atomic/pull.py]: we can probably save another network op here by reworking line 28 (fq_name)

File: Atomic/pull.py (hunk 1 above)
Comment (runcom) [inline Atomic/pull.py]: Did you mean @baude?

Comment (rhatdan) [inline Atomic/pull.py]: Use util.input()

File: Atomic/pull.py (hunk 2 above)
Comment (rhatdan) [inline Atomic/pull.py]: Yes raise ValueError

File: Atomic/pull.py [hunk 4]
```diff
@@ -17,6 +17,8 @@ def cli(subparser):
                        help=_("Specify the storage. Default is currently '%s'.  You can"
                               "change the default by editing /etc/atomic.conf and changing"
                               "the 'default_storage' field." % backend))
+    pullp.add_argument("-y", "--yes", dest="trust", action="store_true",
```
Comment (rhatdan) [inline Atomic/pull.py]: atomic has a flag called --assumeyes, which you should use for this.

```
man atomic
...
       -y --assumeyes
         automatically answer yes for all questions
```

File: Atomic/pull.py [hunk 5]
```diff
@@ -27,6 +29,9 @@ def pull_docker_image(self):
         tag = tag if tag != "" else "latest"
         fq_name = skopeo_inspect("docker://{}".format(self.args.image))['Name']
         image = "docker-daemon:{}:{}".format(fq_name, tag)
+        if get_atomic_config_item(['discover_sigstores'], get_atomic_config()):
+            if not self.discover_sigstore():
```
Comment (rhatdan) [inline Atomic/pull.py]: These should probably combined into a single function.  Put it in self.discover_sigstore()

File: Atomic/pull.py (hunk 1 above)
Comment (rhatdan) [inline Atomic/pull.py]: util.input()

File: Atomic/util.py [hunk 6]
```diff
@@ -247,21 +249,24 @@ def urllib3_disable_warnings():
             if hasattr(urllib3, 'disable_warnings'):
                 urllib3.disable_warnings()
 
-def skopeo_inspect(image, args=None, return_json=True, newline=False):
+def skopeo_inspect(image, args=None, return_json=True, newline=False, fail_silent=False):
```
Comment (rhatdan) [inline Atomic/util.py]: quiet=False

File: Atomic/util.py [hunk 7]
```diff
@@ -381,6 +386,53 @@ def get_atomic_config():
...
         return yaml_load(conf_file)
 
+def write_registry_config(scope):
+    """
+    Write registry sigstore configuration file
+    :param scope: registry string
+    :return: True on success
+    """
+    # FIXME: pending agreement on registry sigstore layout
+    registry_dir = get_atomic_config_item(['registry_sigstore_dir'], get_atomic_config())
```
Comment (rhatdan) [inline Atomic/util.py]: We need to check for errors here or provide a default.

File: Atomic/util.py [hunk 8]
```diff
@@ -381,6 +386,53 @@ def get_atomic_config():
...
+    return False
+
+def install_pubkey(key_name, key_url):
+    """
+    Installs public key to system config directory
+    :param key_name: id of key used as filename
+    :param key_url: download URI of public key
+    :return: pubkey path string or False
+    """
+    pubkeys_dir = get_atomic_config_item(['pubkeys_dir'], get_atomic_config())
```
Comment (rhatdan) [inline Atomic/util.py]: Need a default

**Comment by rhatdan** [comment]
We also want to provide a location in the users homedirs so people can use this for non-privileged containers.

File: Atomic/pull.py (hunk 2 above)
Comment (aweiteka) [inline Atomic/pull.py]: The user is pulling an image. I don't want to exit if we cannot automagically configure trust but I do want to warn them and provide further steps to manually configure trust.

Comment (rhatdan) [inline Atomic/pull.py]: write_err then.  This might need to be controlled by policy?

**Comment by aweiteka** [comment]
This needs to be refactored to use methods from #603 (registries.d configuration) and #604 (policy management)

**Comment by aweiteka** [comment]
@rhatdan @baude please re-review. I plan to add integration tests in another PR to cover this and other signing functionality.

**Comment by rhatdan** [comment]
Needs a rebase.

File: Atomic/trust.py [hunk 9]
```diff
@@ -182,9 +236,93 @@ def write_registry_config_file(self, reg_file, registry, sigstore):
...
         return t[sigstore_type]
+
+    def discover_sigstore(self, pull_image, confirm=None):
+        """
+        Check for registry/repo/sigstore metadata image
+        prompt user for trust on first use workflow
+        :param pull_image: image being pulled, used to discover matching sigstore meta image
+        :param confirm: string
+        """
+        if not util.get_atomic_config_item(['discover_sigstores'], util.get_atomic_config()):
```
Comment (rhatdan) [inline Atomic/trust.py]: Actually can you just pass the assumeyes on the __init__ function.  And refer to it as assumeyes, so we have consistency.

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**